Issue #5057 - Included root context path

gregw opened the pull request:
Root context path in include should be empty string.

gregw committed:
Fixes #5057 Included root context path

Root context path in include should be empty string.

joakime reviewed (approved):
Looks pretty straight forward to me.
+1

lachlan-roberts commented on jetty-servlet/src/test/java/org/eclipse/jetty/servlet/IncludedServletTest.java:
```diff
@@ -173,4 +219,54 @@ private String getPotentialBody(HttpURLConnection connection)
+    public void testIncludeAttributes() throws IOException
+    {
+        URI uri = baseUri.resolve("/attr/one");
+        System.out.println("GET (String): " + uri.toASCIIString());
```
You might want to remove the print here and log instead.

janbartel commented on jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java:
```diff
@@ -591,6 +591,16 @@ public String getContextPathEncoded()
+     * Get the context path in a form suitable to be returned from {@link HttpServletRequest#getContextPath()}
+     * @return Returns the encoded contextPath, or empty string for root context
+     */
+    public String getRequestContextPath()
```
How does this compare to ContextHandler.Context.getContextPath()? see for example https://github.com/eclipse/jetty.project/blob/jetty-9.4.x/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java#L2421.

gregw replied:
Interesting question, because I think `ContextHandler.Context.getContextPath()` is actually wrong in that it should be returning the encoded path. 
Eitherway, I think we should only have one of the two methods.... stand by....

joakime replied:
Question ... 

What happens in this situation ...

``` shell
$ cd /path/to/mybase
$ cd webapps
$ mkdir "this%2Fpath"
$ echo "can you see this" > "this%2Fpath/index.html"
```

What is the context path here?

gregw replied:
I think technically it would be`/this%252fpath` and should be ok other than being confusing.
But we should test and I expect some nasties.

gregw replied:
But I'd prefer not to hold up this PR for that.

janbartel reviewed (changes requested):
Just clarify the difference with ContextHandler.Context.getContextPath() (ie should one use the other?) and then I'm +1.

gregw committed:
Merge branch 'jetty-9.4.x' into jetty-9.4.x-5057-IncludedContextPath

gregw committed:
Issue #5057

merged context path methods as result of review.
ServletContent.getContextPath now returns the encoded contextPath (if anybody is silly enough to have one).
